twine/auth.py

```diff
@@ -0,0 +1,80 @@
+import warnings
+import getpass
+import functools
+
+import keyring
+import munch
+
+from . import utils
+from . import exceptions
+
+
+class Resolver:
+    def __init__(self, settings, username=None, password=None):
+        self.settings = settings
+        self.input = munch.Munch(username=username, password=password)
+
+    @property
+    @functools.lru_cache()
+    def username(self):
+        return utils.get_userpass_value(
+            self.input.username,
+            self.config,
+            key='username',
+            prompt_strategy=self.username_from_keyring_or_prompt,
+        )
+
+    @property
+    @functools.lru_cache()
+    def password(self):
+        return utils.get_userpass_value(
+            self.input.password,
+            self.config,
+            key='password',
+            prompt_strategy=self.password_from_keyring_or_prompt,
+        )
+
+    @property
+    def system(self):
+        return self.settings.system
```

Comment:
This and `config` feel somewhat circular to me, given that `Settings.auth` refers to an instance of `Resolver`. Could `system` and `config` be passed as `__init__` args, instead of `settings`?

Reply:
That's maybe not a bad idea. I was bemoaning this somewhat circular dependency. I am trying to avoid copying state from one object to another (but instead referencing it), but this may be a case where practicality beats purity. I'll experiment.

Reply:
My main objection to passing both `system` and `config` is that `system` is already in `config`, so it makes sense to consolidate those in an object that expresses and makes explicit the relationship between those values. But then I realize that `Settings` already does that, which leads me back to the implementation you see.

Reply:
In the latest implementation, the auth.Resolver expects a simple `config: RepositoryConfig` object.